### PR TITLE
Add error codes duplicates check

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
As asked in #67086.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Fixed formatting issue.

**Comment by Mark-Simulacrum** [comment]
I don't have much time to do review here, so would appreciate a more elaborate PR description that tells me:
 * what this is doing ("duplicates check" is a bit vague, or at least I'm not sure I understand it)
  * in particular, e.g. a description of exactly the conditions intended to be caught, and the cases that we are not intending to catch
 * brief description of what we had before that (may) overlap with this, i.e., other error code checks

**Comment by GuillaumeGomez** [comment]
It counts the occurences of all error code, and detects the duplicates: you're not supposed to be able to use a same error code in two different places (otherwise, we might use one error code for two different errors).

So the check is pretty simple in itself: we read files, count the occurrences and check the numbers.

Before that, it was done through a macro which has been removed since then.

File: src/tools/tidy/src/error_codes_check.rs [hunk 1]
```diff
@@ -117,7 +158,7 @@ pub fn check(path: &Path, bad: &mut bool) {
         eprintln!("{}", err);
     }
     println!("Found {} error codes with no tests", errors.len());
-    if !errors.is_empty() {
+    if !errors.is_empty() || errors_count != 0 {
```
Comment (Mark-Simulacrum) [inline src/tools/tidy/src/error_codes_check.rs]: Can we instead push into the errors vec instead of having two separate modes?

**Review by Mark-Simulacrum — COMMENTED** [review]
Looks like this currently depends on the other PR, too, as we currently have non-stringy error codes.

**Comment by Mark-Simulacrum** [comment]
Do we have consensus as to not having duplicate error code usage? I personally feel like that's a pretty odd policy to have, but if @estebank and @rust-lang/wg-diagnostics feel that we should enforce this then I'm not opposed.

**Comment by oli-obk** [comment]
The policy existed for error codes because it was easy to mix up numbers. If we move to stringy error codes, all we need is a check that each has a test emitting it.

**Comment by Mark-Simulacrum** [comment]
@oli-obk That's only true if we truly stringify error codes (e.g. `borrow-check-use-after-drop`), not the current `"E0303"`, right? (And would need a different parser anyway).

But if mixing up number is the concern, then I guess we can do this. I still feel like that's something that can be fixed whenever it comes up and this doesn't in practice help much, but I also don't usually write errors or diagnostics in general :)

**Comment by oli-obk** [comment]
> But if mixing up number is the concern, then I guess we can do this. I still feel like that's something that can be fixed whenever it comes up and this doesn't in practice help much, but I also don't usually write errors or diagnostics in general :)

We've never had problems with this because we've had this check. I have no idea if it would be a problem in practice ;)

**Comment by Mark-Simulacrum** [comment]
We've not had this check for... some time now, right? In any case, I guess it doesn't matter. I'm fine with landing this (once my comments are resolved), and I probably also need to do another review pass after that to make sure the implementation is doing what it says it is.

**Comment by Mark-Simulacrum** [comment]
I think there was some discussion in Zulip as well, but I don't have time to track that down anyway. The code here is (loosely) waiting on the diagnostics WG anyway probably or blocked on the other PR. Moving to waiting-on-team.

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. reassigning to @oli-obk as member of WG-diagnostics to decide whether to close as unnecessary, or to land.

**Comment by oli-obk** [comment]
Closing as we haven't had this check for quite a while and don't seem to have problems without it.

**Comment by GuillaumeGomez** [comment]
I have to strongly disagree on this opinion: before I migrated to the `librustc_error_code`, we had a check performed at compile-time through a macro. We lost this check and I had already to write fixes about new errors added (with no explanation) which already existed. So definitely, this is needed so I'll reopen the PR.

**Comment by Mark-Simulacrum** [comment]
> I had already to write fixes about new errors added (with no explanation) which already existed

I am confused by this -- I thought this lint was for accidentally reusing the same error code (accidentally), not about not having an error code? I've just re-read your comment https://github.com/rust-lang/rust/pull/68639#issuecomment-580263318 and I feel like it matches my understanding, so it may be helpful to open a new issue or PR instead? Or maybe I'm just misunderstanding...

**Comment by GuillaumeGomez** [comment]
The other lint checks if all error codes have an explanation. This lint checks that each error code is only used once.

**Comment by Centril** [comment]
I am in agreement with @oli-obk and @Mark-Simulacrum. One thing I noted at the meeting is that disallowing reusing an error code can have ungreat architectural implications force you have to have `fn foo_error(..) -> DiagnosticBuilder<'_>` and that can harm the effort towards https://github.com/rust-lang/rust/issues/65031.

**Comment by GuillaumeGomez** [comment]
At this point, I just wonder why I try to maintain/clean up this whole error system. No one wants it and no one wants to make another. That's one problem less off my shoulders. Good luck!

**Pull request closed without merging**